common/src/primitives/height.rs

```diff
@@ -1,18 +1,113 @@
 use parity_scale_codec_derive::{Decode, Encode};
+use std::fmt;
+use std::ops::{Add, Sub};
+
+type HeightIntType = u64;
+type DistanceIntType = i64;
 
 #[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Encode, Decode)]
-pub struct BlockHeight(u64);
+pub struct BlockHeight(HeightIntType);
+
+// Display should be defined for thiserr crate
+impl fmt::Display for BlockHeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<BlockHeight> for HeightIntType {
+    fn from(block_height: BlockHeight) -> HeightIntType {
+        block_height.0
+    }
+}
+
+impl From<HeightIntType> for BlockHeight {
+    fn from(w: HeightIntType) -> BlockHeight {
+        BlockHeight(w)
+    }
+}
+
+impl Add<BlockDistance> for BlockHeight {
+    type Output = Option<BlockHeight>;
+
+    fn add(self, other: BlockDistance) -> Option<BlockHeight> {
+        let height: i64 = self.0.try_into().ok()?;
+        if height < 0 {
+            // we can't do arithmetic on this height anymore. Unless it's a bug, we won't face this in a million years
+            return None;
+        }
+
+        let result = height
+            .checked_add(other.0)
+            .expect("overflow when adding BlockHeight to instant");
+        if result < 0 {
+            return None;
+        }
+
+        let result: u64 = result.try_into().ok()?;
+        Some(Self(result))
+    }
+}
+
+impl Sub<BlockDistance> for BlockHeight {
+    type Output = Option<BlockHeight>;
+
+    fn sub(self, other: BlockDistance) -> Option<BlockHeight> {
+        let h1: i64 = self.0.try_into().ok()?;
+
+        if h1 < 0 {
+            return None;
+        }
+
+        let raw_result: i64 = h1.checked_sub(other.0)?;
+        if raw_result < 0 {
+            return None;
+        }
+
+        let raw_result: u64 = raw_result.try_into().ok()?;
+
+        Some(BlockHeight::new(raw_result))
+    }
+}
+
+impl Add<BlockDistance> for BlockDistance {
+    type Output = Option<BlockDistance>;
+
+    fn add(self, other: BlockDistance) -> Option<BlockDistance> {
+        Some(BlockDistance(self.0.checked_add(other.0)?))
```

Comment:
Again, just a style comment, feel free to ignore (resolve). A more "functional" implementation:
```rust
self.0.checked_add(other.0).map(Self /* or BlockDistance */)
```